File: Leetcode/131. Palindrome Partitioning.py
class Solution:
    def __init__(self):
        self.res=[]
    def isPalindrome(self,t):
        return t==t[::-1]
    def backtrack(self,s:str,temp:List[str],index):
        if index==len(s):
            self.res.append(temp[:])
            return
        t=""
        for i in range(index,len(s)):
            t+=s[i]
            if self.isPalindrome(t):
                temp.append(t)
                self.backtrack(s,temp,i+1)
                temp.pop()
        return 
    def partition(self, s: str) -> List[List[str]]:
        temp=[]
        self.backtrack(s,temp,0)
        return self.res

File: Leetcode/131. Palindrome Partitioning.cpp
class Solution {
public:
    bool isPalindrome(string& s) {
        int l=0,r=s.size()-1;
        while (l < r) {
            if (s[l++] != s[r--]) {
                return false;
            }
        }
        return true;
    }
    vector<vector<string>> res;
    void backtrack(string s,vector<string>& temp,int index){
        if(index==s.size()){
            res.push_back(temp);
            return;
        }
        string t="";
        for(int i=index;i<s.size();i++){
            t+=s[i];
            if(isPalindrome(t)){
                temp.push_back(t);
                backtrack(s,temp,i+1);
                temp.pop_back();
            }
        }
        return;
    }
    vector<vector<string>> partition(string s) {
        vector<string> temp;
        backtrack(s,temp,0);
        return res;
    }
};
